Sources/Features/Breadcrumb/BacktraceBreadcrumbs.swift

```diff
@@ -0,0 +1,128 @@
+
+    public static let none: [BacktraceBreadcrumbType] = []
+}
+
+@objc public enum BacktraceBreadcrumbLevel: Int {
+
+    case debug = 1
+    case info = 2
+    case warning = 3
+    case error = 4
+    case fatal = 5
+
+    var description: String {
+        switch self {
+        case .debug:
+            return "debug"
+        case .info:
+            return "info"
+        case .warning:
+            return "warning"
+        case .error:
+            return "error"
+        case .fatal:
+            return "fatal"
+        }
+    }
+}
+
+@objc public class BacktraceBreadcrumbs: NSObject {
+
+    private var breadcrumbsLogManager: BacktraceBreadcrumbsLogManager?
+    private var backtraceNotificationObserver: BacktraceNotificationObserver?
+    private var breadcrumbLevel: BacktraceBreadcrumbLevel?
+    private var breadcrumbTypes: [BacktraceBreadcrumbType]?
+    private(set) var isBreadcrumbsEnabled: Bool = false
+
+    public func enableBreadcrumbs(_ breadcrumbSettings: BacktraceBreadcrumbSettings = BacktraceBreadcrumbSettings()) {
+        do {
+            self.breadcrumbLevel = breadcrumbSettings.breadcrumbLevel
+            self.breadcrumbTypes = breadcrumbSettings.breadcrumbTypes
+            self.breadcrumbsLogManager = try BacktraceBreadcrumbsLogManager(breadcrumbSettings: breadcrumbSettings)
+            if breadcrumbSettings.breadcrumbTypes.contains(where: { $0.rawValue == BacktraceBreadcrumbType.system.rawValue }) {
+                backtraceNotificationObserver = BacktraceNotificationObserver(breadcrumbs: self)
+                backtraceNotificationObserver?.enableNotificationObserver()
+            }
+
+            try BreadcrumbsInfo.breadcrumbFile = breadcrumbSettings.getBreadcrumbLogPath()
+
+            isBreadcrumbsEnabled = true
+            _ = addBreadcrumb("Breadcrumbs enabled.")
+        } catch {
+            BacktraceLogger.warning("\(error.localizedDescription) \nWhen enabling breadcrumbs, breadcrumbs is disabled")
+            disableBreadcrumbs()
+        }
+    }
+
+    public func disableBreadcrumbs() {
+        self.isBreadcrumbsEnabled = false
+        self.backtraceNotificationObserver = nil
+        self.breadcrumbsLogManager = nil
+        self.breadcrumbTypes = nil
+        self.breadcrumbLevel = nil
+
+        // Remove breadcrumbs attachment
+        BreadcrumbsInfo.breadcrumbFile = nil
+
+        // Remove currentBreadcrumbsId, which prevents it from being added
+        BreadcrumbsInfo.currentBreadcrumbsId = nil
+    }
+
+    func addBreadcrumb(_ message: String,
+                       attributes: [String: String]? = nil,
+                       type: BacktraceBreadcrumbType = BacktraceBreadcrumbType.manual,
+                       level: BacktraceBreadcrumbLevel = BacktraceBreadcrumbLevel.info) -> Bool {
+        if let breadcrumbsLogManager = breadcrumbsLogManager, allowBreadcrumbsToAdd(level) {
```

Comment:
what if the breadcrumbsLogManager is disabled because the enable method threw an internal error and we disabled breadcrumbs? We got a null reference exception here.

Reply:
Actually Swift is really cool and prevents that via optional unwrapping. If it's null it falls through to `return false`

```
Unwrap an optional type with optional binding
Optional binding is similar to using an if-else block. The only subtle difference is that if the optional value is not nil, the unwrapped value is assigned to a new constant and further operations are performed on the constant.

You can do this using the if-let statement:

var password:String? = "$tr0ngp@$$w0rd"

if let unwrappedpass = password {
	print("Password is \(unwrappedpass)") //Password is $tr0ngp@$$w0rd
}
```
https://www.freecodecamp.org/news/optional-types-in-swift/